model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java

```diff
@@ -472,24 +472,23 @@ public Long getGroupsCountByNameContaining(RealmModel realm, String search) {
 
     @Override
     public Stream<GroupModel> getTopLevelGroupsStream(RealmModel realm) {
-        RealmEntity ref = em.getReference(RealmEntity.class, realm.getId());
-
-        return ref.getGroups().stream()
-                .filter(g -> GroupEntity.TOP_PARENT_ID.equals(g.getParentId()))
-                .map(g -> session.groups().getGroupById(realm, g.getId()))
-                .sorted(Comparator.comparing(GroupModel::getName));
+        return getTopLevelGroupsStream(realm, null, null);
     }
 
     @Override
     public Stream<GroupModel> getTopLevelGroupsStream(RealmModel realm, Integer first, Integer max) {
-        Stream<String> groupIds =  em.createNamedQuery("getTopLevelGroupIds", String.class)
+        TypedQuery<String> groupsQuery =  em.createNamedQuery("getTopLevelGroupIds", String.class)
                 .setParameter("realm", realm.getId())
-                .setParameter("parent", GroupEntity.TOP_PARENT_ID)
-                .setFirstResult(first)
-                .setMaxResults(max)
-                .getResultStream();
+                .setParameter("parent", GroupEntity.TOP_PARENT_ID);
 
-        return closing(groupIds.map(realm::getGroupById));
+        if (first != null && max != null) {
```

Comment:
This something we should probably unify across the whole codebase because I saw more approaches to this [(for example).](https://github.com/keycloak/keycloak/blob/master/model/jpa/src/main/java/org/keycloak/models/jpa/JpaRealmProvider.java#L460). Maybe we should add this to the cleanup task.

I did it this way because it is used here, but I will fix: https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/services/resources/admin/GroupsResource.java#L84

Reply:
Absolutely agree, we should follow single approach to these parameters. Thanks for fixing.